### PR TITLE
Replace codefresh with GitHub actions

### PR DESCRIPTION
**Pull request by dylanbannon**
Unresolved issues:

- [x] Cloudpossebot needs to relax with the Codefresh builds.
- [x] I had to make some very hacky changes to the `cloudposse/docs/content` folder to get the `smoketest` target to pass. They may not be ideal in terms of communicating with the public. Is this a @Gowiem issue?
- [x] The `Push Site to S3`, `Push New Index to Algolia`, and `Update Cloudflare Cache` steps have not been tested yet.

### CODE REVIEW

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        cat artifacts.json
+        jq --version
+        jq '[.["artifacts"][] | {name: .name, created: .created_at} | select(.name=="htmltest-cache")] | length' artifacts.json > number_of_caches.txt
+        export number_of_caches=`cat number_of_caches.txt`
+        if [ ${number_of_caches} > 0 ]; then
+            echo "::set-output name=cache-exists::true"
+            curl -H "Accept: application/vnd.github.v3+json" https://api.github.com/repos/cloudposse/docs/actions/artifacts | jq '[.["artifacts"][] | select(.name=="htmltest-cache")] | sort_by(.created_at) | reverse[0].archive_download_url' > artifact_url.txt
+            export artifact_url=`cat artifact_url.txt`
+            echo "::set-output name=cache-url::${artifact_url}"
+        fi
```
Comment (osterman) [inline .github/workflows/build.yaml:64]: This is a lot of logic and probably redflag again we're doing something wrong.

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    - name: "Download Hugo Site Artifact, If Existent"
+      if:  ${{ steps.artifact_test.outputs.cache-exists }}
+      run: |
+        curl -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" ${{ steps.artifact_test.outputs.cache-url }}
+        ls -hratl
+    - name: "(Re-)Generate cache"
+      run: |
```
Comment (osterman) [inline .github/workflows/build.yaml:72]:
```suggestion
    - name: "(Re-)Generate cache"
      if: succeeded() || failed()
      run: |
```

File: .github/workflows/build.yaml (hunk 1 above)
Comment (osterman) [reply .github/workflows/build.yaml:64]: I suggest just removing all this and using either:

```
if: succeeded() || failed()
```
or 

```
continue-on-error: true
```

File: Makefile [hunk 3]
```diff
@@ -27,7 +27,7 @@ export DOCKER_IMAGE_NAME ?= $(DOCKER_IMAGE):$(DOCKER_TAG)
 export DOCKER_BUILD_FLAGS =
 
 ifeq ($(wildcard /.dockerenv),)
-export DOCKER_RUN := docker run -it --rm -v $(CURDIR):/src -p $(HUGO_PORT):$(HUGO_PORT) -e YARN_BUILD_DISABLED -e GITHUB_BASIC_AUTH $(DOCKER_IMAGE_NAME)
```
Comment (osterman) [inline Makefile:30]: So, locally, we'll want to have it be interactive (so we can hit `^C`). 

Let' instead add `DOCKER_RUN_FLAGS`

File: Makefile [hunk 4]
```diff
@@ -27,7 +27,7 @@ export DOCKER_IMAGE_NAME ?= $(DOCKER_IMAGE):$(DOCKER_TAG)
 export DOCKER_BUILD_FLAGS =
```
Comment (osterman) [inline Makefile:28]:
```suggestion
export DOCKER_BUILD_FLAGS =
export DOCKER_RUN_FLAGS ?= -it --rm
```

File: Makefile [hunk 5]
```diff
@@ -27,7 +27,7 @@ export DOCKER_IMAGE_NAME ?= $(DOCKER_IMAGE):$(DOCKER_TAG)
...
 
 ifeq ($(wildcard /.dockerenv),)
-export DOCKER_RUN := docker run -it --rm -v $(CURDIR):/src -p $(HUGO_PORT):$(HUGO_PORT) -e YARN_BUILD_DISABLED -e GITHUB_BASIC_AUTH $(DOCKER_IMAGE_NAME)
+export DOCKER_RUN := docker run --rm -v $(CURDIR):/src -p $(HUGO_PORT):$(HUGO_PORT) -e YARN_BUILD_DISABLED -e GITHUB_BASIC_AUTH $(DOCKER_IMAGE_NAME)
```
Comment (osterman) [inline Makefile:30]:
```suggestion
export DOCKER_RUN := docker run $(DOCKER_RUN_FLAGS) -v $(CURDIR):/src -p $(HUGO_PORT):$(HUGO_PORT) -e YARN_BUILD_DISABLED -e GITHUB_BASIC_AUTH $(DOCKER_IMAGE_NAME)
```

File: .github/workflows/build.yaml [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    - created
+
+env:
+  MASTER_BRANCH: master
+  BUILD_HARNESS_VERSION: 0.5.5
+  HUGO_CONFIG: deploy.toml
+  HTMLTEST_CONFIG: .htmltest.codefresh.yml
+  #TMPDIR: /codefresh/volume
+  YARN_BUILD_DISABLED: false
+  COMPONENTS_BUILD: true
```
Comment (osterman) [inline .github/workflows/build.yaml:15]:
```suggestion
  COMPONENTS_BUILD: true
  DOCKER_RUN_FLAGS: "--rm"
```

File: .github/workflows/build.yaml [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        curl -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" ${{ steps.artifact_test.outputs.cache-url }}
+        ls -hratl
+    - name: "(Re-)Generate cache"
+      run: |
+        make real-clean smoketest
+    - name: "Upload Hugo Site Artifact"
+      uses: actions/upload-artifact@v2
+      with:
+        name: htmltest-cache
+        path: ./public
```
Comment (osterman) [inline .github/workflows/build.yaml:78]: Actually, we don't want to cache this. We want to cache the `.htmltest` folder or whatever `OutputDir` is set to.

File: .github/workflows/build.yaml [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    - name: "Set up Docker Buildx"
+      uses: docker/setup-buildx-action@v1
+    - name: "Build docker image"
+      uses: docker/build-push-action@v2
+      with:
+        tags: cloudposse/docs
+        outputs: type=docker,dest=/tmp/image.tar
+    - name: "Import Image"
+      run: |
+        docker load --input /tmp/image.tar
```
Comment (osterman) [inline .github/workflows/build.yaml:33]: Why do we have the image.tar?

File: .github/workflows/build.yaml [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        make real-clean hugo/build
+    - name: "Verify that Hugo files were built"
+      run: |
+        ls -lhatr
+        ls -lhatr ./public
+
+  hydrate-cache:
+    name: "Hydrate Cache"
+    runs-on: ubuntu-latest
+    steps:
```
Comment (osterman) [inline .github/workflows/build.yaml:48]:
```suggestion
```

File: .github/workflows/build.yaml [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+            curl -H "Accept: application/vnd.github.v3+json" https://api.github.com/repos/cloudposse/docs/actions/artifacts | jq '[.["artifacts"][] | select(.name=="htmltest-cache")] | sort_by(.created_at) | reverse[0].archive_download_url' > artifact_url.txt
+            export artifact_url=`cat artifact_url.txt`
+            echo "::set-output name=cache-url::${artifact_url}"
+        fi
+    - name: "Download Hugo Site Artifact, If Existent"
+      if:  ${{ steps.artifact_test.outputs.cache-exists }}
+      run: |
+        ls -hratl
+        curl -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" ${{ steps.artifact_test.outputs.cache-url }}
+        ls -hratl
```
Comment (osterman) [inline .github/workflows/build.yaml:70]:
```suggestion
    - name: Get branch
      if: (github.event_name == 'release' && github.event.action == 'created') || github.event.pull_request.head.repo.full_name == github.repository
      id: prepare
      run: |
  
        if [[ $GITHUB_REF == refs/tags/* ]]; then
          VERSION=${GITHUB_REF#refs/tags/}
        elif [[ $GITHUB_REF == refs/pull/* ]]; then
          VERSION=pr-${{ github.event.pull_request.number }}-merge
        fi
      
        echo ::set-output name=version::${VERSION}
   - name: Pull htmltest-cache-branch
      id: htmltest-cache-branch
      uses: actions/cache@v2
      with:
        path: .htmltest
        key: htmltest-cache-${{ steps.prepare.outputs.version }}

    - name: Fetch cache from default branch, if no cache hit
      if: steps.htmltest-cache-branch.outputs.cache-hit != 'true'
      uses: actions/cache@v2
      with:
        path: .htmltest
        key: htmltest-cache-master
```

File: .github/workflows/build.yaml [hunk 11]
```diff
@@ -0,0 +1,78 @@
+on:
```
Comment (osterman) [inline .github/workflows/build.yaml:1]:
```suggestion
on:
  push:
    branches:
    - master
```

File: .github/workflows/build.yaml (hunk 7 above)
Comment (osterman) [inline .github/workflows/build.yaml:78]:
```suggestion
  - name: "Push htmltest-cache for branch"
    uses: actions/cache@v2
    with:
      path: | 
        .htmltest/ 
      key: htmltest-cache-${{ steps.prepare.outputs.version }}
```

File: .github/workflows/build.yaml [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+  release:
+    types:
+    - created
+
+env:
+  MASTER_BRANCH: master
+  BUILD_HARNESS_VERSION: 0.5.5
+  HUGO_CONFIG: deploy.toml
+  HTMLTEST_CONFIG: .htmltest.codefresh.yml
+  #TMPDIR: /codefresh/volume
```
Comment (osterman) [inline .github/workflows/build.yaml:13]:
```suggestion
  TMPDIR: ${{ github.workspace }}
```

File: .github/workflows/build.yaml (hunk 7 above)
Comment (dylanbannon) [reply .github/workflows/build.yaml]: The previous cache pull attempts will push the cache automatically upon successful completion of the job, if cache-hit==false.

**Comment by dylanbannon** [comment]
`actions/cache` will only refresh a cache if it fails to hit it earlier in the same job. This means that we can't both pull an existing cache and push a newer version of the same cache.

There's a 3rd party fix for this: https://github.com/actions/cache/pull/498. Thoughts?

**Comment by osterman** [comment]
new workflow psuedo code

```
jobs:
  - hugo-build
    - run: make release for smoketest
    - run: make release for release 
    - run: upload artifact of static site
  - hugo-cache
    - run:   donwload artifact of static site
    - run: make smoketest || true (wihtout building site)
    - run: make set cache (push the cache)
  - hugo-validate
    - run: download artifact of static site 
    - run: use cache
    - run: make smoketest (without building site) and error on failure
  - hugo-release:
     - run: download artifact of static site   
     - run: make deploy to s3 when release 
 ```

File: .github/workflows/build.yaml [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+        else
+          echo "::set-output name=previous_sha_short::false"
+        fi
+    - name: "Get cache location"
+      id: get_cache_location
+      run: |
+        [[ "$(cat ${HTMLTEST_CONFIG} | grep OutputDir)" =~ \"(.+)\" ]] && export location="${BASH_REMATCH[1]}"
+        echo "Cache location ${location}"
+        echo "Present directory: $(pwd)"
+        echo "${location}" > ./cache_location.txt
```
Comment (osterman) [inline .github/workflows/build.yaml:71]: Why do we need to do this?

File: .github/workflows/build.yaml [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+    steps:
+    - name: "Get SHA"
+      id: get_sha
+      run: |
+        echo "::set-output name=current_sha_short::$(git rev-parse --short HEAD)"
+    - name: "Download cache location artifact"
+      uses: actions/download-artifact@v2
+      with:
+        name: cache-location
+        path: ./cache_location.txt
```
Comment (osterman) [inline .github/workflows/build.yaml:109]: The cache location seems deterministic to me

File: .github/workflows/build.yaml [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+    - name: "Upload Hugo site artifact"
+      uses: actions/upload-artifact@v2
+      with:
+        name: hugo-site
+        path: ./public
+    #hugo-cache-pull:
+    - name: "Get SHAs"
+      id: get_shas
+      run: |
+        echo "::set-output name=current_sha_short::$(git rev-parse --short HEAD)"
```
Comment (osterman) [inline .github/workflows/build.yaml:58]: Why are we not using the github envs?

**Comment by dylanbannon** [comment]
Lingering issues:
1. I'm unsure how to disable the Codefresh builds.
2. I'm unsure whether the `if` condition attached to the `hugo-deploy` job is appropriate.

File: .github/workflows/build.yaml (hunk 14 above)
Comment (dylanbannon) [reply .github/workflows/build.yaml]: yeah, you're right. we're using an env now.

File: .github/workflows/build.yaml (hunk 15 above)
Comment (dylanbannon) [reply .github/workflows/build.yaml]: https://stackoverflow.com/questions/59810838/how-to-get-the-short-sha-for-the-github-workflow

For computing `current_sha_short` it seems just as easy to stick with the way it is as it is to parse `github.sha`.

For computing `previous_sha_short`, though, adding the `fetch-depth: 2` parameter to `actions:checkout@v2` and then `git rev-parse`ing that is the only option.

If you think it'd make things simpler, both `current_sha_short` and `previous_sha_short` could be computed once and then passed around as artifacts, as needed.

**Comment by dylanbannon** [comment]
Lingering issues:

1. I'm unsure how to disable the Codefresh builds.
2. I'm unsure whether the if condition attached to the hugo-deploy job is appropriate.

**Comment by dylanbannon** [comment]
All tests will pass on pushes to an open pull request!

Now, issues to address during review:

1. Cloudpossebot needs to relax with the Codefresh builds.
2. The `if` condition attached to the `hugo-deploy` job in `.github/workflows/build.yaml` may not trigger on all the desired deployment cases. Right now, it only triggers when a release gets tagged.
3. I had to make some very hacky changes to the `cloudposse/docs/content` folder to get the `smoketest` target to pass. They may not be ideal in terms of communicating with the public.
4. The `hugo-deploy` job has not been tested yet, since I'm not sure how to mock up the `Push Site to S3`, `Push New Index to Algolia`, and `Update Cloudflare Cache` steps.

**Comment by osterman** [comment]
@dylanbannon please rebase


image

File: .github/workflows/build.yaml [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+      uses: actions/download-artifact@v2
+      with:
+        name: docker-image
+        path: /tmp/
+    - name: "Import Image"
+      run: docker load --input /tmp/image.tar
+    - name: "Extract cache pull location"
+      id: get_cache_pull_location
+      run: |
+        export cache_pull_location=$( "${{ env.CACHE_LOCATION }}" | sed 's|\(.*\)/.*|\1|')
```
Comment (osterman) [inline .github/workflows/build.yaml:81]: isn't this just the `dirname`? e.g. `dirname ${{env.CACHE_LOCATION }}`

File: .github/workflows/build.yaml [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+        echo "::set-output name=cache-pull-location::cache_pull_location"
+    - name: "Pull previous htmltest cache cache, if it exists"
+      id: htmltest-cache
+      uses: actions/cache@v2
+      with:
+        path: ${{ steps.get_cache_pull_location.outputs.cache-pull-location }}
+        key: htmltest-cache-${{ steps.get_sha.outputs.previous_sha_short }}
+    - name: "(Re-)Generate htmltest cache"
+      run: make real-clean smoketest || true
```
Comment (osterman) [inline .github/workflows/build.yaml:90]:
```suggestion
    - name: "Update htmltest cache (ignore failures)"
      run: make real-clean smoketest || true
```

File: .github/workflows/build.yaml [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+        path: ./public
+    - name: "Pull new htmltest cache"
+      id: htmltest-cache
+      uses: actions/cache@v2
+      with:
+        path: ${{ env.CACHE_LOCATION }}
+        key: htmltest-cache-${{ steps.get_sha.outputs.current_sha_short }}
+    - name: "Push Site to S3"
+      run: |
+        set +e -x
```
Comment (osterman) [inline .github/workflows/build.yaml:188]: Instead set the defaults

See https://github.com/cloudposse/actions/blob/master/.github/workflows/test-command.yml#L6-L9

File: .github/workflows/build.yaml [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+        make deploy
+    - name: "Push New Index to Algolia"
+      run: |
+        set +e -x
```
Comment (osterman) [inline .github/workflows/build.yaml:192]: Use defaults

File: .github/workflows/build.yaml [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+        make reindex
+    - name: "Update Cloudflare Cache"
+      run: |
+        set +e -x
```
Comment (osterman) [inline .github/workflows/build.yaml:196]: Use defaults

File: .github/workflows/build.yaml (hunk 16 above)
Comment (dylanbannon) [reply .github/workflows/build.yaml:81]: Wow, that is a much simpler way of doing that.

**Comment by dylanbannon** [comment]
I've `rebase`d the commit history of this branch into oblivion, and GitHub is still saying that 353 files were changed. When I run `git diff master..replace-codefresh-with-github-actions`, though, I only get output for a small number of files.

File: .github/workflows/build.yaml (hunk 15 above)
Comment (dylanbannon) [reply .github/workflows/build.yaml]: Well, I ended up doing exactly this.

File: .github/workflows/build.yaml [hunk 21]
```diff
@@ -0,0 +1,330 @@
+on:
```
Comment (osterman) [inline .github/workflows/build.yaml:1]:
```suggestion
name: hugo
on:
```

File: content/tools/hugo.md [hunk 22]
```diff
@@ -13,5 +13,4 @@ What we like about it is that it's written in Go, speaks "markdown" and uses Go-
 
 Want to host your own static site? Here's how we do it.
 
-1. Our [`terraform-root-module/aws/docs`](https://github.com/cloudposse/terraform-root-modules/tree/master/aws/docs) reference architecture describes how we provision the S3 bucket along with CloudFront CDN.
-2. Our [`codefresh.yml`](https://github.com/cloudposse/docs/blob/master/codefresh.yml) shows how we CI/CD our documentation and deploy it to S3 for all tagged releases.
+1. Our [`codefresh.yml`](https://github.com/cloudposse/docs/blob/master/codefresh.yml) shows how we CI/CD our documentation and deploy it to S3 for all tagged releases.
```
Comment (osterman) [inline content/tools/hugo.md:16]:
```suggestion
1. Our [`codefresh.yml`](https://github.com/cloudposse/docs/tree/master/.github) shows how we CI/CD our documentation and deploy it to S3 for all tagged releases.
```

**Review by osterman — APPROVED** [review]

**Pull request merged**